compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp

```diff
@@ -27,6 +28,170 @@ namespace Flow {
+      ArrayAttr::get(linalgOp->getContext(), newIndexingMaps),
+      linalgOp.getIteratorTypesAttr(), linalgOp.getDocAttr(),
+      linalgOp.getLibraryCallAttr(), bodyBuilder);
+
+  rewriter.replaceOp(linalgOp, newLinalgOp.getResults());
+
+  return success();
+}
+
+static LogicalResult tryRaiseToCollapseShape(AffineMap inputIndexingMap,
+                                             AffineMap outputIndexingMap,
+                                             Value input, Value output,
+                                             linalg::GenericOp linalgOp,
+                                             RewriterBase &rewriter) {
+  // Output shape must be smaller than input shape.
+  if (outputIndexingMap.getNumResults() >= inputIndexingMap.getNumResults())
+    return failure();
+  // Output map should be identity.
+  if (!outputIndexingMap.isIdentity())
+    return failure();
+
+  // Try to build a reassociation. The input indexing map should be the output
+  // indexing map, with some dimensions collapsed. Collapsed dimensions
+  // are represented by 0.
+  SmallVector<ReassociationIndices> reassociation;
+  ReassociationIndices curr;
+  unsigned currOutDim = 0;
+  for (auto [idx, expr] : llvm::enumerate(inputIndexingMap.getResults())) {
+    if (expr == outputIndexingMap.getResult(currOutDim)) {
+      curr.push_back(idx);
+      reassociation.push_back(curr);
+      curr.clear();
+      ++currOutDim;
+      continue;
+    }
+    if (expr.isa<AffineConstantExpr>() &&
+        expr.cast<AffineConstantExpr>().getValue() == 0) {
+      curr.push_back(idx);
+      continue;
+    }
+    return failure();
+  }
+  if (currOutDim != outputIndexingMap.getNumResults())
+    return failure();
+
+  // Push the remaining to the last reassociation if it exists.
+  if (!curr.empty())
+    reassociation.back().append(curr.begin(), curr.end());
+
+  rewriter.replaceOpWithNewOp<tensor::CollapseShapeOp>(linalgOp, input,
+                                                       reassociation);
+  return success();
+}
+
+static LogicalResult tryRaiseToView(linalg::GenericOp linalgOp,
+                                    RewriterBase &rewriter) {
+  if (!linalgOp.hasTensorSemantics())
+    return failure();
+
+  // Assume there is only 1 input, and 1 init tensor.
+  if (linalgOp.getNumDpsInputs() != 1 || linalgOp.getNumDpsInits() != 1)
+    return failure();
+  OpOperand *inputOperand = linalgOp.getDpsInputOperand(0);
+  OpOperand *outputOperand = linalgOp.getDpsInitOperand(0);
+
+  // Check if linalg.yield yields a block arguement.
+  auto yieldOp = dyn_cast<linalg::YieldOp>(linalgOp.getBody()->getTerminator());
+  if (!yieldOp)
+    return failure();
+  // Check if the result is a block argument.
+  auto blockArg = dyn_cast<BlockArgument>(yieldOp.getOperand(0));
+  if (!blockArg)
+    return failure();
+  // Check if the block argument is an argument of the linalgOp.
+  if (!llvm::is_contained(linalgOp.getBody()->getArguments(), blockArg))
```

Comment:
you should be able to do `blockArg.getOwner() == linalgOp.getBody()`? You also need to check that the `blockArg.getArgumentNumber() == 0`